python/cuml/utils/input_utils.py

```diff
@@ -0,0 +1,177 @@
+    """
+    Returns ctype pointer of a cudf column
+    """
+    return cudf.bindings.cudf_cpp.get_column_data_ptr(col._column)
+
+
+def get_dtype(X):
+    """
+    Returns dtype of obj as a Numpy style dtype (like np.float32)
+    """
+    if isinstance(X, cudf.DataFrame):
+        dtype = np.dtype(X[X.columns[0]]._column.dtype)
+    elif (isinstance(X, cudf.Series)):
+        dtype = np.dtype(X._column.dtype)
+    elif isinstance(X, np.ndarray):
+        dtype = X.dtype
+    elif cuda.is_cuda_array(X):
+        dtype = X.dtype
+    elif cuda.devicearray.is_cuda_ndarray(X):
+        dtype = X.dtype
+
+    return dtype
+
+
+def input_to_array(X, order='F', deepcopy=False,
+                   check_dtype=False, convert_to_dtype=False,
+                   check_cols=False, check_rows=False):
+    """
+    Convert input X to device array suitable for C++ methods
+    Acceptable input formats:
+    * cuDF Dataframe - returns a deep copy always
+    * cuDF Series - returns by reference or a deep copy depending on
+        `deepcopy`
+    * Numpy array - returns a copy in device always
+    * cuda array interface compliant array (like Cupy) - returns a
+        reference unless deepcopy=True
+    * numba device array - returns a reference unless deepcopy=True
+
+    Returns a new device array if the input was not a numba device array.
+    Returns a reference to the input X if its a numba device array or cuda
+        array interface compliant (like cupy)
+    """
+
+    if convert_to_dtype:
+        X = convert_dtype(X, to_dtype=convert_to_dtype)
+        check_dtype = False
+
+    if isinstance(X, cudf.DataFrame):
+        dtype = np.dtype(X[X.columns[0]]._column.dtype)
+        if order == 'F':
+            X_m = X.as_gpu_matrix(order='F')
+        elif order == 'C':
+            X_m = cuml.utils.numba_utils.row_matrix(X)
+
+    elif (isinstance(X, cudf.Series)):
+        if deepcopy:
+            X_m = X.to_gpu_array()
+        else:
+            X_m = X._column._data.mem
+
+    elif isinstance(X, np.ndarray):
+        dtype = X.dtype
+        X_m = rmm.to_device(np.array(X, order=order, copy=False))
+
+    elif cuda.is_cuda_array(X):
+        # Use cuda array interface to create a device array by reference
+        X_m = cuda.as_cuda_array(X)
+
+    elif cuda.devicearray.is_cuda_ndarray(X):
+        X_m = X
+
+    else:
+        msg = "X matrix format " + str(X.__class__) + " not supported"
+        raise TypeError(msg)
+
+    dtype = X_m.dtype
+
+    if check_dtype:
+        if dtype != check_dtype:
+            del X_m
+            raise TypeError("ba")
+
+    n_rows = X_m.shape[0]
+    if len(X_m.shape) > 1:
+        n_cols = X_m.shape[1]
+    else:
+        n_cols = 1
+
+    if check_cols:
+        if n_cols != check_cols:
+            raise ValueError("ba")
+
+    if check_rows:
+        if n_rows != check_rows:
+            raise ValueError("ba")
+
+    X_ptr = get_dev_array_ptr(X_m)
+
+    return X_m, X_ptr, n_rows, n_cols, dtype
```

Comment:
(Maybe controversial, maybe a bad idea, but...) We could consider returning this as a NamedTuple like `namedtuple('DeviceArray', 'X_m X_ptr n_rows n_cols dtype')`. Seems like that has few downsides since callers can just treat it as a plain tuple, but you could also reference by name for clarity when needed.

Reply:
Also, what is the class of `X_m` and is it always consistent? Does it already include n_rows, n_cols, dtype information? (looks like it includes dtype at least)

Reply:
`X_m` is always a numba device array indeed since I create it. I wanted to leave the return values without being a tuple to avoid any conflic with calling the function from cython and assigning it to a `uintptr_t`, will try to change it to see if it works and how it looks, though I kind of like the simplicity of just returning the values instead of a tuple. Can always be convinced otherwise of course :P